sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -384,17 +385,34 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     val bis = new ByteArrayInputStream(bytes)
     val ins = new DataInputStream(codec.compressedInputStream(bis))
 
-    new Iterator[InternalRow] {
+    new NextIterator[InternalRow] {
+      private val taskContext: TaskContext = TaskContext.get()
+      if (taskContext != null) {
+        taskContext.addTaskCompletionListener[Unit] { _ => closeIfNeeded() }
+      }
```

Comment:
Shall we simply like the following?
```scala
- private val taskContext: TaskContext = TaskContext.get()
- if (taskContext != null) {
-   taskContext.addTaskCompletionListener[Unit] { _ => closeIfNeeded() }
- }
+ Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => closeIfNeeded()))
```

Reply:
BTW, shall we declare this before `new NextIterator[InternalRow] {` instead of here?

Reply:
The `closeIfNeeded()` method is a member of NextIterator.